File: src/types/importV2Presentation.ts
import type { AgentKind } from "./agent";
import type { AgentAssistanceTrigger } from "./importV2Agent";
import type { CapabilityRequirement } from "./importV2File";
import type { BackendTask } from "./task";
import type {
  MigrationStatus,
} from "./importV2Migration";

export type ImportWorkbenchSection = "workbench" | "capabilities" | "history";
export type ImportWorkbenchQueueFilter = "all" | "active" | "ready" | "needs_action" | "failed";

export interface ImportWorkbenchPreferences {
  schemaVersion: 1;
  activeSection: ImportWorkbenchSection;
  queueFilter: ImportWorkbenchQueueFilter;
  workbenchScrollTop: number;
  capabilitiesScrollTop: number;
  historyScrollTop: number;
  sourceMethodsExpanded: boolean;
}

export interface ImportWorkbenchPreferencesRequest {
  projectId: string;
  projectRootPath: string;
}

export interface SaveImportWorkbenchPreferencesRequest extends ImportWorkbenchPreferencesRequest {
  preferences: ImportWorkbenchPreferences;
}

export interface ImportPreviewContent {
  sessionId: string;
  itemId: string;
  candidateId: string | null;
  title: string;
  markdown: string;
  truncated: boolean;
  totalBytes: number;
  sha256: string;
  target?: ImportPreviewTarget;
  quality?: import("./importV2").QualityReport;
  rawLabel?: string;
  resources?: ImportPreviewResource[];
  comparison?: ImportPreviewComparison | null;
}

export interface ImportPreviewComparison {
  currentMarkdown: string;
  mergedMarkdown?: string | null;
}

export interface ImportPreviewTarget {
  disposition: "new_source" | "update" | "merge" | "duplicate";
  sourceId: string | null;
  versionId: string | null;
  wikiPath: string | null;
}

export interface ImportPreviewResource {
  source: string;
  name: string;
  kind: "image" | "attachment" | "subtitle" | "transcript" | "metadata" | "source_evidence" | "source_snapshot" | "markdown";
  sizeBytes: number;
  dataUrl: string | null;
}
export interface ImportFrontendReadiness {
  backendVersion: string;
  active: boolean;
  migrationStatus: MigrationStatus;
  unfinishedSessionId: string | null;
  legacyHistoryAvailable: boolean;
  files?: ImportFeatureReadiness[];
  platforms?: ImportPlatformReadiness[];
  abilities?: ImportFeatureReadiness[];
  capabilities?: ImportCapabilityReadiness[];
}

export interface GetImportRestrictedContentStatusV2Request {
  projectId: string;
  projectRootPath: string;
}

export interface ImportRestrictedContentStatus {
  confirmationRequired: boolean;
}

export interface ImportPlatformReadiness {
  id: string;
  available: boolean;
  reasonCode?: string | null;
}

export type ImportHistoryAction = "open_detail" | "open_result" | "view_logs" | "update_wiki";

export interface GetImportPreviewContentV2Request {
  projectId: string;
  projectRootPath: string;
  sessionId: string;
  itemId: string;
  candidateId: string | null;
  historyBatchId?: string | null;
}

export interface GetImportFrontendReadinessV2Request {
  projectId: string;
  projectRootPath: string;
}

export interface ImportHistoryEntry {
  id: string;
  title: string;
  status: string;
  sessionId: string | null;
  batchId: string | null;
  taskId: string | null;
  startedAt: string | null;
  updatedAt: string | null;
  completedAt: string | null;
  legacyReadOnly: boolean;
  itemIds: string[];
  availableActions: ImportHistoryAction[];
  snapshotAvailable?: boolean;
}

export function canOpenHistoricalResult(entry: Pick<ImportHistoryEntry, "status" | "availableActions">): boolean {
  return (entry.status === "completed" || entry.status === "partially_committed")
    && entry.availableActions.includes("open_result");
}

export interface LegacyHistoryEntry {
  id: string;
  title: string;
  status: string;
  startedAt: string | null;
  updatedAt: string | null;
  completedAt: string | null;
  evidencePath: string;
  legacyReadOnly: true;
  availableActions: string[];
  canRetry: false;
  canDelete: false;
  canReplaceSource: false;
}

export interface LegacyHistoryWarning {
  code: string;
  message: string;
  evidencePath: string;
}

export interface ImportHistoryPage {
  entries: ImportHistoryEntry[];
  legacyReadOnly: LegacyHistoryEntry[];
  nextCursor: string | null;
  warnings: LegacyHistoryWarning[];
}

export interface ListImportHistoryV2Request {
  projectId: string;
  projectRootPath: string;
  cursor: string | null;
  limit?: number;
}

export interface ImportCapabilityRequirement {
  requirement: CapabilityRequirement;
  route: string;
  available: boolean;
  installable: boolean;
  compressedBytes: number | null;
  installedBytes: number | null;
  modelBytes: number | null;
  license: string | null;
  fallback: string | null;
}

export interface GetImportCapabilityRequirementV2Request {
  projectId: string;
  projectRootPath: string;
  sessionId: string;
  itemId: string;
}

export interface InstallImportCapabilityV2Request {
  projectId: string;
  projectRootPath: string;
  sessionId: string;
  itemId: string;
  capabilityId: string;
  acknowledgeInstall: boolean;
}

export interface GetImportAsrEnablementPlanV2Request {
  projectId: string;
  projectRootPath: string;
  sessionId: string;
  itemId: string;
}

export type ImportAsrDependencyKind = "media_runtime" | "engine" | "model" | "language_support";

export interface ImportAsrDependency {
  kind: ImportAsrDependencyKind;
  name: string;
  available: boolean;
  bundledWithCapability: boolean;
  source: string;
  license: string;
}

export interface ImportAsrProfilePlan {
  profile: import("./importV2").ImportAsrProfile;
  capabilityId: string;
  engineName: string;
  modelName: string;
  available: boolean;
  installable: boolean;
  downloadBytes: number | null;
  installedBytes: number | null;
  modelBytes: number | null;
  device: string;
  estimatedSeconds: number | null;
  unavailableReasonCode: string | null;
  dependencies: ImportAsrDependency[];
}

export interface ImportAsrEnablementPlan {
  recommendedProfile: import("./importV2").ImportAsrProfile;
  availableMemoryBytes: number | null;
  availableDiskBytes: number | null;
  mediaDurationSeconds: number | null;
  installLocation: string | null;
  localOnly: boolean;
  profiles: ImportAsrProfilePlan[];
}

export interface BeginImportLoginV2Request {
  projectId: string;
  projectRootPath: string;
  sessionId: string;
  itemId: string;
  platform: string;
}

export interface ConnectorSessionRef {
  sessionId: string;
  platform: string;
  state: string;
  accountSummary?: string | null;
  lastVerifiedAt?: string | null;
}

export interface CompleteImportLoginV2Request {
  projectId: string;
  projectRootPath: string;
  importSessionId: string;
  itemId: string;
  connectorSessionId: string;
}

export interface CompleteImportLoginResult {
  connectorSession: ConnectorSessionRef;
  resumedItemIds: string[];
  tasks: BackendTask[];
}

export interface RevokeImportLoginV2Request {
  sessionId: string;
  platform?: string | null;
}

export interface ImportFeatureReadiness {
  id: string;
  available: boolean;
  reasonCode?: string | null;
}

export interface ImportCapabilityReadiness {
  capabilityId: string;
  route: string;
  available: boolean;
  reasonCode?: string | null;
}

export interface AuthorizeImportPrivateTargetV2Request {
  projectId: string;
  projectRootPath: string;
  sessionId: string;
  itemId: string;
  url: string;
}

export interface AuthorizeLocalAsrV2Request {
  projectId: string;
  projectRootPath: string;
  sessionId: string;
  itemId: string;
  profile: import("./importV2").ImportAsrProfile;
  language?: string | null;
}
export type AuthorizeBilibiliAsrV2Request = AuthorizeLocalAsrV2Request;

export interface StartImportAgentAssistanceV2Request {
  projectId: string;
  projectRootPath: string;
  sessionId: string;
  itemId: string;
  trigger: AgentAssistanceTrigger;
  agentKind: AgentKind;
}

File: src-tauri/src/models/import_v2_presentation.rs
use serde::{Deserialize, Serialize};

use super::import_v2::{ImportAsrProfile, QualityReport};
use super::import_v2_file::CapabilityRequirement;
use super::import_v2_migration::{LegacyHistoryEntry, LegacyHistoryWarning, MigrationStatus};

pub const IMPORT_V2_PREVIEW_MAX_BYTES: u64 = 2 * 1024 * 1024;
pub const IMPORT_V2_WORKBENCH_PREFERENCES_PATH: &str = ".app/import-workbench.json";

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum ImportWorkbenchSection {
    #[default]
    Workbench,
    Capabilities,
    History,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum ImportQueuePreference {
    #[default]
    All,
    Active,
    Ready,
    NeedsAction,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportWorkbenchPreferences {
    pub schema_version: u32,
    pub active_section: ImportWorkbenchSection,
    pub queue_filter: ImportQueuePreference,
    pub workbench_scroll_top: u32,
    pub capabilities_scroll_top: u32,
    pub history_scroll_top: u32,
    pub source_methods_expanded: bool,
}

impl Default for ImportWorkbenchPreferences {
    fn default() -> Self {
        Self {
            schema_version: 1,
            active_section: ImportWorkbenchSection::Workbench,
            queue_filter: ImportQueuePreference::All,
            workbench_scroll_top: 0,
            capabilities_scroll_top: 0,
            history_scroll_top: 0,
            source_methods_expanded: true,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportWorkbenchPreferencesRequest {
    pub project_id: String,
    pub project_root_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SaveImportWorkbenchPreferencesRequest {
    pub project_id: String,
    pub project_root_path: String,
    pub preferences: ImportWorkbenchPreferences,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct GetImportPreviewContentV2Request {
    pub project_id: String,
    pub project_root_path: String,
    pub session_id: String,
    pub item_id: String,
    pub candidate_id: Option<String>,
    #[serde(default)]
    pub history_batch_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ImportPreviewContent {
    pub session_id: String,
    pub item_id: String,
    pub candidate_id: Option<String>,
    pub title: String,
    pub markdown: String,
    pub truncated: bool,
    pub total_bytes: u64,
    pub sha256: String,
    pub target: ImportPreviewTarget,
    pub quality: QualityReport,
    pub raw_label: String,
    #[serde(default)]
    pub resources: Vec<ImportPreviewResource>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub comparison: Option<ImportPreviewComparison>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportPreviewComparison {
    pub current_markdown: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub merged_markdown: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportPreviewTarget {
    pub disposition: String,
    pub source_id: Option<String>,
    pub version_id: Option<String>,
    pub wiki_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportPreviewResource {
    pub source: String,
    pub name: String,
    pub kind: String,
    pub size_bytes: u64,
    pub data_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct GetImportFrontendReadinessV2Request {
    pub project_id: String,
    pub project_root_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportFrontendReadiness {
    pub backend_version: String,
    pub active: bool,
    pub migration_status: MigrationStatus,
    pub unfinished_session_id: Option<String>,
    pub legacy_history_available: bool,
    #[serde(default)]
    pub files: Vec<ImportFeatureReadiness>,
    #[serde(default)]
    pub platforms: Vec<ImportPlatformReadiness>,
    #[serde(default)]
    pub abilities: Vec<ImportFeatureReadiness>,
    #[serde(default)]
    pub capabilities: Vec<ImportCapabilityReadiness>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportPlatformReadiness {
    pub id: String,
    pub available: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason_code: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportFeatureReadiness {
    pub id: String,
    pub available: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason_code: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportCapabilityReadiness {
    pub capability_id: String,
    pub route: String,
    pub available: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason_code: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ListImportHistoryV2Request {
    pub project_id: String,
    pub project_root_path: String,
    pub cursor: Option<String>,
    pub limit: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ImportHistoryAction {
    OpenDetail,
    OpenResult,
    ViewLogs,
    UpdateWiki,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportHistoryEntry {
    pub id: String,
    pub title: String,
    pub status: String,
    pub session_id: Option<String>,
    pub batch_id: Option<String>,
    pub task_id: Option<String>,
    pub started_at: Option<String>,
    pub updated_at: Option<String>,
    pub completed_at: Option<String>,
    pub legacy_read_only: bool,
    pub item_count: u64,
    pub committed_count: u64,
    pub failed_count: u64,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub sample_labels: Vec<String>,
    pub available_actions: Vec<ImportHistoryAction>,
    /// False for pre-snapshot records that can only be shown through a
    /// best-effort live-session compatibility fallback.
    #[serde(default)]
    pub snapshot_available: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "camelCase")]
pub struct ImportHistoryPage {
    pub entries: Vec<ImportHistoryEntry>,
    pub legacy_read_only: Vec<LegacyHistoryEntry>,
    pub next_cursor: Option<String>,
    pub warnings: Vec<LegacyHistoryWarning>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct GetImportHistoryDetailV2Request {
    pub project_id: String,
    pub project_root_path: String,
    pub batch_id: String,
    pub cursor: Option<String>,
    pub limit: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RebuildImportHistoryIndexV2Request {
    pub project_id: String,
    pub project_root_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ImportHistoryDetailPage {
    pub entry: ImportHistoryEntry,
    pub items: Vec<crate::models::import_v2::ImportItem>,
    pub next_cursor: Option<String>,
    pub total: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct GetImportCapabilityRequirementV2Request {
    pub project_id: String,
    pub project_root_path: String,
    pub session_id: String,
    pub item_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportCapabilityRequirement {
    pub requirement: CapabilityRequirement,
    pub route: String,
    pub available: bool,
    pub installable: bool,
    pub compressed_bytes: Option<u64>,
    pub installed_bytes: Option<u64>,
    pub model_bytes: Option<u64>,
    pub license: Option<String>,
    pub fallback: Option<String>,
    pub requirement_revision: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct InstallImportCapabilityV2Request {
    pub project_id: String,
    pub project_root_path: String,
    pub session_id: String,
    pub item_id: String,
    pub capability_id: String,
    pub requirement_revision: String,
    pub acknowledge_install: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub asr_profile: Option<ImportAsrProfile>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub recognition_language: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct GetImportAsrEnablementPlanV2Request {
    pub project_id: String,
    pub project_root_path: String,
    pub session_id: String,
    pub item_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ImportAsrDependencyKind {
    MediaRuntime,
    Engine,
    Model,
    LanguageSupport,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportAsrDependency {
    pub kind: ImportAsrDependencyKind,
    pub name: String,
    pub available: bool,
    pub bundled_with_capability: bool,
    pub source: String,
    pub license: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportAsrProfilePlan {
    pub profile: ImportAsrProfile,
    pub capability_id: String,
    pub engine_name: String,
    pub model_name: String,
    pub available: bool,
    pub installable: bool,
    pub download_bytes: Option<u64>,
    pub installed_bytes: Option<u64>,
    pub model_bytes: Option<u64>,
    pub device: String,
    pub estimated_seconds: Option<u64>,
    pub unavailable_reason_code: Option<String>,
    pub dependencies: Vec<ImportAsrDependency>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportAsrEnablementPlan {
    pub requirement_revision: String,
    pub recommended_profile: ImportAsrProfile,
    pub available_memory_bytes: Option<u64>,
    pub available_disk_bytes: Option<u64>,
    pub media_duration_seconds: Option<u64>,
    pub install_location: Option<String>,
    pub local_only: bool,
    pub profiles: Vec<ImportAsrProfilePlan>,
}
